SQL: JDBC: fix access to the Manifest for non-entry JAR URLs

bpintea opened the pull request:
The JDBC driver will attempt to read its version from the Manifest file
embedded into its JAR. The URL pointing to the JAR can be provided in a
few ways.

So far, accessing the Manfiest was attempted by getting a URLConnection
out of the URL and then getting an input stream out of this connection.
For file JAR URLs, this only works however if the URL points to the
driver as a JAR file entry (i.e. `<sub-url>!/jdbc-driver.jar!/`). If
that's not the case, the JarURLConnection will throw an IOException.

This PR fixes that: in case the URL points to a JAR entry
(`jar:file:<path>/jdbc-driver.jar!/`), the manifest is read directly with
`JarURLConnection#getManifest()`.

Fixes #56759.

bpintea committed:
JDBC: fix access to the Manifest for non-entry JAR

The JDBC driver will attempt to read its version from the Manifest file
embedded into its JAR. The URL pointing to the JAR can be provided in a
few ways.

So far, accessing the Manfiest was attempted by getting a URLConnection
out of the URL and then getting an input stream out of this connection.
For file JAR URLs, this only works however if the URL points to the
driver as a JAR file entry (i.e. <sub-url>!/jdbc-driver.jar!/). If
that's not the case, the JarURLConnection will throw an IOException.

This commit fixes that: in case the URL points to a JAR entry
(jar:file:<path>/jdbc-driver.jar!/), the manifest is read directly with
JarURLConnection#getManifest().

bpintea committed:
Fix style errors

Remove unused import.

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

astefan commented on x-pack/plugin/sql/sql-client/src/test/java/org/elasticsearch/xpack/sql/client/VersionTests.java:
```diff
@@ -31,27 +31,74 @@ public void testInvalidVersion() {
 
         // create the jdbc driver file
-        try (JarOutputStream jdbc = new JarOutputStream(Files.newOutputStream(innerJarPath, StandardOpenOption.CREATE), jdbcJarManifest)) {}
+        try (JarOutputStream __ = new JarOutputStream(Files.newOutputStream(jarPath, StandardOpenOption.CREATE), jdbcJarManifest)) {}
```
:-)

bpintea replied:
Yeh, `_` is now a keyword (besides a common-ish practice to name unused vars :-) ). I just wanted to have the line within the line len limit and no extra line added.

astefan commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -71,23 +72,36 @@
+                    try (JarInputStream jar = new JarInputStream(conn.getInputStream())) {
+                        manifest = jar.getManifest();
+                    }
+                } while (false);
```
`do { } while (false)` or use a function with the same break conditions (but using `return` to "jump" at the end)?

costin replied:
This section needs improving.
the `do {} while (false)` is unnecessary. The break can also be removed by checking if the manifest is still null before the try block:

```
if (url.getProtocol()...) {
}

if (manifest == null) {
 try (JarInputStream...)
}
```

Further more, if the entry is not-null what would be the returned manifest?
```
if (url.getProtocol().equals("jar")) {
   JarURLConnection jarConn = (JarURLConnection) conn;
   manifest = jarConn.getManifest(); <-- ? what would this return?

```

bpintea replied:
> do { } while (false)

I thought this might raise some eyebrows. :-)

> or use a function with the same break conditions (but using return to "jump" at the end)?

I've turned the frame loop into a function. Hope this is what you were suggesting.

> the do {} while (false) is unnecessary.

Generally this avoids an unnecessary fetch&compare. 
But in this case if the manifest is still `null` (shouldn't, but if), `conn.getInputStream()` would raise an exception (vs. just returning the "null version").

> if the entry is not-null what would be the returned manifest?

If the entry is not null, it'd simply return the [outermost] JAR's manifest (since the entry can be any class, i.e. non a JAR). Let me know if I still didn't get your question, though.

astefan commented on x-pack/plugin/sql/sql-client/src/test/java/org/elasticsearch/xpack/sql/client/VersionTests.java:
```diff
@@ -31,27 +31,74 @@ public void testInvalidVersion() {
+    }
+
+    public void testVersionFromJar() throws IOException {
+        byte[] parts = {1, 2, 3};
```
Why is this `{1, 2, 3}` and the previous test is `{0, 1, 2}`. It is relevant for these tests to have different versions to test? If not, and for the sake of using whatever versions, can you use a random value?

bpintea replied:
Right, randomising it is a better idea. I've added that, thanks.

astefan reviewed (commented):
Looks good. Appreciate the quick fix for the issue. Left few cosmetic comments.

matriv commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -71,23 +72,36 @@
         if (urlStr.endsWith(".jar") || urlStr.endsWith(".jar!/")) {
             try {
                 URLConnection conn = url.openConnection();
-                conn.setUseCaches(false);
```
I think there was a reason for that, but can't remember. @astefan do you ?

astefan replied:
I didn't notice this change when I reviewed this. Good point.
I _think_ I added that because the temporary jars that were created during the test couldn't be cleaned-up properly. And I would keep the `setUseCaches(false)` as is. I was getting something like this:

```
java.io.IOException: Could not remove the following files (in the order of attempts):
   C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001\tempDir-002\foo.jar: java.nio.file.FileSystemException: C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001\tempDir-002\foo.jar: The process cannot access the file because it is being used by another process.
   C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001\tempDir-002: java.nio.file.DirectoryNotEmptyException: C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001\tempDir-002
   C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001: java.nio.file.DirectoryNotEmptyException: C:\Users\AppData\Local\Temp\org.elasticsearch.xpack.sql.client.VersionTests_123F282E30B0C58F-001
	at __randomizedtesting.SeedInfo.seed([123F282E30B0C58F]:0)
	at org.apache.lucene.util.IOUtils.rm(IOUtils.java:319)
	at org.apache.lucene.util.TestRuleTemporaryFilesCleanup.afterAlways(TestRuleTemporaryFilesCleanup.java:216)
	at com.carrotsearch.randomizedtesting.rules.TestRuleAdapter$1.afterAlways(TestRuleAdapter.java:31)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

costin replied:
Avoid file locking, at least on Windows.

costin replied:
Not sure why my previous comment doesn't show up but it's because it avoids file locking on Windows as it frees the stream early on.

bpintea replied:
Good catch, thanks @matriv. It was unintentional. I've put it back.

matriv reviewed (commented):
LGTM, left a comment about `setUseCaches(false)`

costin reviewed (approved):
Left some comments regarding styling but looks good otherwise, thanks for the fast turnaround!

bpintea committed:
Address review comments

- turn frame loop into function call
- make version randomized
- readd conn caching disabling, unintentionally removed.

matriv reviewed (approved):
LGTM

bpintea commented:
@elasticmachine run elasticsearch-ci/bwc

bpintea commented:
@elasticmachine run elasticsearch-ci/default-distro

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/jdbc_version_read
